### PR TITLE
ELEC-449: Fix up telemetry

### PR DESCRIPTION
**Pull request by ckitagawa**
This PR is large and does a lot of things but they are so interlinked that it is difficult to break into smaller parts.

- Add Main Pack, Aux and DCDC values to the Display
- Correctly Maps IDs
- Properly writes to a database on supported systems or a file otherwise
- Tests cross compiling for RPI
- Fixes up requestAnimationFrame to actually end. (Huge CPU savings)
- Fixes parsing logic to parse based on schema protobufs and uses a dynamic map to allow JSON access by field name in websocket
- Update to modern CSS
- Fix `tsc` config

Remaining work:

- Make Travis build and deploy releases
- Configure the PI to execute the auto start script
- Properly integrate with codegen-tooling via update script. At present just copied and pasted from codegen.

### CODE REVIEW

File: client/src/ts/dd_main.ts [hunk 1]
```diff
@@ -3,6 +3,17 @@ import dial = require('./dial');
 import readout = require('./readout');
 import vis = require('./visibility');
 import animate = require('./animate');
+import canDefs = require('./can_msg_defs');
+
+// Helpers
+function toFixedTrunc(value: number, n: number): string {
```
Comment (tituschow) [inline client/src/ts/dd_main.ts]: The fact that this is necessary bothers me. You can't use `.toPrecision()`?

File: pkg/db/can_db.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+	"sync"
+
+	log "github.com/golang/glog"
+	_ "github.com/mattn/go-sqlite3" // DB impl
+
+	"telemetry/pkg/msgs"
+	"telemetry/pkg/pubsub"
+)
+
+// RunDb creates a table and sinks all "CAN" pubsub messages into it via the WirteMsg method.
```
Comment (tituschow) [inline pkg/db/can_db.go]: s/WirteMsg/WriteMsg/

File: pkg/db/sql_api.go [hunk 3]
```diff
@@ -30,7 +48,9 @@ func WriteMsg(db *sql.DB, msg msgs.CAN) error {
 func TimeWindowedRead(db *sql.DB, canID uint16, from time.Time, to time.Time) ([]msgs.CAN, error) {
 	rows, err := db.Query(`
     SELECT
+      can.source,
       can.id,
+      can.rtr,
```
Comment (tituschow) [inline pkg/db/sql_api.go]: We don't actually use the RTR bit anywhere. It would be more useful to store the DLC?

File: pkg/db/sql_api.go [hunk 4]
```diff
@@ -30,7 +48,9 @@ func WriteMsg(db *sql.DB, msg msgs.CAN) error {
 func TimeWindowedRead(db *sql.DB, canID uint16, from time.Time, to time.Time) ([]msgs.CAN, error) {
 	rows, err := db.Query(`
     SELECT
+      can.source,
       can.id,
```
Comment (tituschow) [inline pkg/db/sql_api.go]: Is this the msg ID or raw CAN ID?

File: pkg/msgs/can_msgs.go [hunk 5]
```diff
@@ -1,17 +1,158 @@
...
+
+func parse(rawID uint32, rawData uint64, canMsg *CAN) {
+	canMsg.Timestamp = uint64(time.Now().UnixNano()) / uint64(time.Millisecond)
+	canMsg.Source = uint8(rawID & 0xF)
+	canMsg.RTR = ((rawID & 0x10) >> 4) == 1
+	canMsg.ID = uint16((rawID & 0x7E0) >> 5)
+	srcMsg := findCanMessage(uint32(canMsg.ID))
+	if srcMsg == nil {
+		log.Errorf("Error: no matching schema for", canMsg.ID)
+		canMsg.Data["raw"] = rawData
```
Comment (tituschow) [inline pkg/msgs/can_msgs.go]: We should store the DLC for raw data scenarios

File: pkg/msgs/can_msgs.go [hunk 6]
```diff
@@ -1,17 +1,158 @@
...
+func NewCAN(rawID uint32, rawData uint64) CAN {
+	var canMsg CAN
+	canMsg.Data = make(map[string]interface{})
+	parse(rawID, rawData, &canMsg)
+	return canMsg
+}
+
+func parse(rawID uint32, rawData uint64, canMsg *CAN) {
+	canMsg.Timestamp = uint64(time.Now().UnixNano()) / uint64(time.Millisecond)
+	canMsg.Source = uint8(rawID & 0xF)
```
Comment (tituschow) [inline pkg/msgs/can_msgs.go]: Is there a struct or something that we can use instead of magic masks and shifts?

File: pkg/msgs/can_msgs.go [hunk 7]
```diff
@@ -1,17 +1,158 @@
...
+	canMsg.ID = uint16((rawID & 0x7E0) >> 5)
+	srcMsg := findCanMessage(uint32(canMsg.ID))
+	if srcMsg == nil {
+		log.Errorf("Error: no matching schema for", canMsg.ID)
+		canMsg.Data["raw"] = rawData
+	}
+	switch srcMsg.GetCanData().GetFrame().(type) {
+	case *canpb.CanData_U8:
+		data := srcMsg.GetCanData().GetU8()
+		name := data.GetFieldName_1()
```
Comment (tituschow) [inline pkg/msgs/can_msgs.go]: Too bad this can't be an array or something

File: pkg/sources/serial/serial.go [hunk 8]
```diff
@@ -46,15 +46,17 @@ func Run(port string, bus *pubsub.MessageBus) {
...
 			}
 		} else {
 			// Parse the input and store it into a canPacket. Note that this ignores
 			// the header and the newline.
 			packet := canPacket{}
 			binary.Read(bytes.NewBuffer(buf), binary.LittleEndian, &packet)
-
-			bus.Publish("CAN", msgs.NewCAN(packet.ID, packet.Data))
+			hdr := packet.Header & 0xFFFFFF
+			if hdr == 0x585443 || hdr == 0x585243 {
```
Comment (tituschow) [inline pkg/sources/serial/serial.go]: Probably want to replace this magic

Comment (ckitagawa) [reply pkg/sources/serial/serial.go]: Done.

File: pkg/msgs/can_msgs.go (hunk 7 above)
Comment (ckitagawa) [reply pkg/msgs/can_msgs.go]: Yeah, but such is the nature of Go. That's why there is often a lot of codegen done for Go.

File: pkg/msgs/can_msgs.go (hunk 6 above)
Comment (ckitagawa) [reply pkg/msgs/can_msgs.go]: No, Go has no support for bitfields.

File: pkg/msgs/can_msgs.go (hunk 5 above)
Comment (ckitagawa) [reply pkg/msgs/can_msgs.go]: Done.

File: pkg/db/sql_api.go (hunk 4 above)
Comment (ckitagawa) [reply pkg/db/sql_api.go]: Msg ID. I assume that we are using our network layer for CAN messages.

File: pkg/db/sql_api.go (hunk 3 above)
Comment (ckitagawa) [reply pkg/db/sql_api.go]: Done. I'll just store both.

File: pkg/db/can_db.go (hunk 2 above)
Comment (ckitagawa) [reply pkg/db/can_db.go]: Done.

File: client/src/ts/dd_main.ts (hunk 1 above)
Comment (ckitagawa) [reply client/src/ts/dd_main.ts]: Done.

**Pull request merged**